lib/dry/system/auto_registrar.rb

```diff
@@ -26,37 +27,40 @@ def initialize(container)
 
       # @api private
       def finalize!
-        Array(config.auto_register).each { |dir| call(dir) }
+        config.component_dirs.each do |dir|
+          call(dir.path) if dir.auto_register
+        end
       end
 
       # @api private
       def call(dir)
         registration_config = Configuration.new
+
         yield(registration_config) if block_given?
+
         components(dir).each do |component|
-          next if !component.auto_register? || registration_config.exclude.(component)
+          next unless register_component?(component, registration_config)
 
-          container.require_component(component) do
-            register(component.identifier, memoize: registration_config.memoize) {
-              registration_config.instance.(component)
-            }
-          end
+          container.register(component.identifier, memoize: registration_config.memoize) {
+            registration_config.instance.(component)
+          }
         end
       end
 
       private
 
-      # @api private
       def components(dir)
-        files(dir)
-          .map { |file_name| [file_name, file_options(file_name)] }
-          .map { |file_name, options| component(relative_path(dir, file_name), **options) }
-          .reject { |component| registered?(component.identifier) }
+        files(dir).map { |file_name|
+          Component.new(
+            relative_path(dir, file_name),
+            file_path: file_name,
+            **MagicCommentsParser.(file_name)
+          )
+        }
```

Comment:
By calling `Component.new` here we no longer initialize the component through `Container.component`, which is doing the extra work of trying to locate the component again across all the configured component_dirs. That's not necessary at all here, since we've already crawled a directory and have a known file path for each component! This change alone may have a measurable positive impact on the auto-registration phase for large projects!